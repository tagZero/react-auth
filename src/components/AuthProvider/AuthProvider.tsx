import React, { createContext, useContext, useState } from 'react';
import { useHistory } from 'react-router-dom';
import { AuthServiceType, NotificationState } from '../AuthService/AuthService.type';

const AuthContext = createContext({} as AuthServiceType);
export const useAuth = () => useContext(AuthContext);

const AuthProvider: React.FC<{ service: AuthServiceType }> = ({ service, children }) => {
  const [authData, setAuthData] = useState();
  const [notification, setNotification] = useState<NotificationState>();
  const history = useHistory();

  const {
    login,
    logout,
    register,
    resetPassword,
    isAuthenticated,
    modules,
    options,
    getModulePath
  } = service;

  service.setNotification = setNotification;
  service.setAuthData = setAuthData;
  service.history = history;

  return (
    <AuthContext.Provider
      value={{
        authData,
        isAuthenticated,
        login,
        logout,
        register,
        resetPassword,
        modules,
        options,
        getModulePath,
        notification
      }}
    >
      {children}
    </AuthContext.Provider>
  );
};

export default AuthProvider;
