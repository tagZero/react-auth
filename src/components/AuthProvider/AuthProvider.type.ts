export interface LoginPropsType {
  email: string;
  password: string;
}

export interface RegisterPropsType extends LoginPropsType {
  firstName: string;
  lastName: string;
  birthDate: number;
}

export interface AuthProviderPropsType {
  isAuthenticated?: () => boolean;
  loadingState?: boolean;
  login?: (props: LoginPropsType, context?: AuthProviderType) => Promise<any>;
  logout?: (context?: AuthProviderType) => Promise<any>;
  register?: (props: RegisterPropsType, context?: AuthProviderType) => Promise<any>;
  resetPassword?: (props: { email: string }, context?: AuthProviderType) => Promise<any>;
  modules?: any;
  options?: any;
}

export interface AuthProviderType extends AuthProviderPropsType {
  loading: boolean;
  getModulePath: (module: string) => string;
  setLoading?: (props: any) => any;
  setModules?: (props: any) => any;
  setOptions?: (props: any) => any;
  setToken?: (token: string | null) => any;
  token?: string | null;
}
