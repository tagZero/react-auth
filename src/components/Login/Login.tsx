import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import { useAuth } from '../AuthProvider/AuthProvider';
import { LoginPropsType } from '../AuthProvider/AuthProvider.type';

const Login = () => {
  const [notification, setNotification] = useState<any>();
  const context = useAuth();
  const { login, modules, getModulePath, options } = context;

  const onSubmit = async (event) => {
    event.preventDefault();
    const keys = ['email', 'password'];
    const props = keys.reduce(
      (obj, key) => ({ ...obj, [key]: event.target[key].value }),
      {}
    ) as LoginPropsType;

    try {
      await login(props, context);
      setNotification({ type: 'success', message: modules.login.successMessage });
    } catch (err) {
      setNotification({ type: 'error', message: err });
    }
  };

  const logoStyle = options.logoUrl ? { backgroundImage: `url(${options.logoUrl}` } : null;

  return (
    <main className="auth-container">
      <div className="auth-logo" style={logoStyle} />
      {options.logoTitle ? <div className="auth-title">{options.logoTitle}</div> : null}
      <div className="auth-form-container auth-tile">
        {notification?.type === 'error' ? (
          <div className="auth-form-error">{notification.message}</div>
        ) : null}
        {notification?.type === 'success' ? (
          <div className="auth-form-success">{notification.message}</div>
        ) : null}
        <div className="auth-form-title">{modules.login.title}</div>
        <p className="auth-form-description">{modules.login.description}</p>
        <form className="auth-form" onSubmit={onSubmit}>
          <div className="auth-form-row">
            <input type="email" name="email" id="email" placeholder="E-mail" required />
          </div>
          <div className="auth-form-row">
            <input type="password" name="password" id="password" placeholder="Password" required />
          </div>
          <div className="auth-form-submit">
            <input type="submit" value="Submit" />
          </div>
          <ul className="auth-form-link">
            <li>
              <Link to={getModulePath('register')}>{modules.register.text}</Link>
            </li>
            <li>
              <Link to={getModulePath('resetPassword')}>{modules.resetPassword.text}</Link>
            </li>
          </ul>
        </form>
      </div>
    </main>
  );
};

export default Login;
