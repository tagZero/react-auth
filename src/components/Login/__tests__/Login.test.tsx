import React from 'react';
import { MemoryRouter as Router } from 'react-router-dom';
import { render } from '@testing-library/react';
import AuthProvider from '../../AuthProvider/AuthProvider';
import Login from '../Login';

describe('Login', () => {
  const tree = () =>
    render(
      <Router>
        <AuthProvider>
          <Login />
        </AuthProvider>
      </Router>
    );

  test('renders properly', () => {
    const { asFragment } = tree();
    expect(asFragment()).toMatchSnapshot();
  });
});
