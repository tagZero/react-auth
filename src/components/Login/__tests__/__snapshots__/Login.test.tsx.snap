// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders properly 1`] = `
<DocumentFragment>
  <main
    class="auth-container"
  >
    <div
      class="auth-logo"
    />
    <div
      class="auth-form-container auth-tile"
    >
      <div
        class="auth-form-title"
      >
        Login
      </div>
      <p
        class="auth-form-description"
      >
        Please provide your credentials below to login.
      </p>
      <form
        class="auth-form"
      >
        <div
          class="auth-form-row"
        >
          <input
            id="email"
            name="email"
            placeholder="E-mail"
            required=""
            type="email"
          />
        </div>
        <div
          class="auth-form-row"
        >
          <input
            id="password"
            name="password"
            placeholder="Password"
            required=""
            type="password"
          />
        </div>
        <div
          class="auth-form-submit"
        >
          <input
            type="submit"
            value="Submit"
          />
        </div>
        <ul
          class="auth-form-link"
        >
          <li>
            <a
              href="/auth/register"
            >
              Create new account
            </a>
          </li>
          <li>
            <a
              href="/auth/reset-password"
            >
              Forgot password?
            </a>
          </li>
        </ul>
      </form>
    </div>
  </main>
</DocumentFragment>
`;
